### ISSUE TITLE
Make primitive parsers generic over input to improve streaming support

### ISSUE THREAD

**Issue by demurgos** [Geal/nom#1053]
This is issue is a feature request to improve streaming support.

# Context

I maintain [an SWF parser](https://github.com/open-flash/swf-parser/) written with nom 5. So far, the parser expected the content to be completely loaded before parsing it. This is a major issue as it is expected to support SWF files streamed through the network: I want to add streaming support. Ideally it would even allow to read the content of the SWF file without even having to hold its bytes in memory (but only the parsed content).

I want to explore different strategies to handle data buffering for my parser but I am currently restricted by the primitive parsers exposed by nom to handle the numbers: they require the input to be a `&[u8]`. This in turn means that the input must be contiguous in memory.

The three strategies I want to compare are:
- Append-only buffer: allocate a `Vec<u8>` with enough memory to hold the whole uncompressed file (the SWF format makes it easy to predict) and append chunks to it as they become available.
- `Vec<u8>` that only keeps the unparsed data: occasionally drop the already parsed input at the start of the array with something like `Vec::split_off`. It requires copy/move the unparsed data at the end of the vec.
- Use a ring buffer backed by a `Dequeue` so it is possible to remove the parsed data without having to move/copy the unparsed data.

When memory is not a concern, I expect the first strategy to be the fastest. It is also the simplest.
When trying to only keep in memory the least amount of data required, I want to compare the performance of the second and third implementation.
The issue is that a `Dequeue` stores its data using two slices, not one. By configuring virtual memory, [`slice_deque`](https://crates.io/crates/slice_deque) enables to view these these slice as contiguous, but this crate does not support all targets (for example `wasm32-unknown-unknown`).

# Issue

**I would like to passe my input as `(&[u8], &[u8])`** and compare the result on performance: is a continuous input with copies better than a discontinuous input without copies?
I suspect that the latter is faster but I can't check it at the moment because my parser can only accept `&[u8]` because it calls to `nom::number::streaming` functions which require `&[u8]`. **I would like these primitive parsers to become generic over their input.**

# Proof of concept

`nom` already defines enough traits to make it possible so I checked what it might look like with `le_u16`.

The current streaming `le_u16` implementation looks like this:

```rust
pub fn le_u16<'a, E: ParseError<&'a [u8]>>(i: &'a[u8]) -> IResult<&'a[u8], u16, E> {
  if i.len() < 2 {
    Err(Err::Incomplete(Needed::Size(2)))
  } else {
    let res = ((i[1] as u16) << 8) + i[0] as u16;
    Ok((&i[2..], res))
  }
}
```

It compiles to a branch, simple read and return.

Using the available `nom` traits, we can get the following generic implementation:
```rust
pub fn le_u16<I, E>(i: I) -> IResult<I, u16, E>
  where
    I: InputLength,
    I: Slice<RangeFrom<usize>>,
    I: InputIter<Item=u8>,
    E: ParseError<I>,
{
  if i.input_len() < 2 {
    Err(Err::Incomplete(Needed::Size(2)))
  } else {
    let res: u16 = i.iter_indices().take(2).fold(0u16, |acc, (idx, x)| acc + (u16::from(x) << (8 * idx)));
    Ok((i.slice(2..), res))
  }
}
```

These traits abstract over `&[u8]` by basically representing it as an [ExactSizeIterator](https://doc.rust-lang.org/core/iter/trait.ExactSizeIterator.html) iterator over `u8`.
I consider the code above to be idiomatic Rust. Unfortunately it requires the nightly to compiler to be compiled as efficiently as the current implementation when monomorphized with `&[u8]`. Using the stable compiler, it introduces extra jumps or multiple single-byte reads (instead of a single 2-byte read).

It's possible to get a generic implementation to be as efficient using some non-idiomatic code to compute `res`:
```rust
    let mut it: I::IterElem = i.iter_elem();
    let b0: u8 = it.next().unwrap();
    let b1: u8 = it.next().unwrap();
    let res: u16 = b0 as u16 + ((b1 as u16) << 8);
```

You can [check these implementations with the Compiler Explorer](https://godbolt.org/z/eUQ9tG).

# Conclusion

In conclusion, I believe that it should be possible to make the primitive parsers generic over their input, but maintaining performance for the `&[u8]` case requires to use non-idiomatic code or a recent compiler version (so it would probably have to be behind a feature flag until its stable).
Making it generic would allow the parser to accept other inputs such as `(&[u8], &[u8])` (enabling the use of ring buffers backed by a `Deque`) or even an `ExactSizeIterator<Item=u8>`.

# Related issues

- https://github.com/Geal/nom/issues/1028 seems somewhat related. Supporting generic input would allow to use a `BufReader` directly after implementing traits for it (but it may require additional info such as the total length).

**Comment by demurgos** [comment]
A few weeks ago I spent some time to update the primitive parsers to be generic. The change turned out to be much larger than expected and pretty complex. I don't remain the details, but the main issue was that it required to define additional traits and propagate them. It also required a non-idiomatic style to keep the same performance.

Ultimately, I also decided that I don't need to use a circular buffer for my input so I no longer need this feature that much. I still think that it would be nice to have. I feel that waiting for native support for const generics should help with the implementation.

**Comment by Geal** [comment]
it was still a valid point to require that those pasers be generic, so it was one in f2d4ab1
